encodings/fastlanes/src/for/compute.rs

```diff
@@ -73,25 +75,68 @@ impl SliceFn for FoRArray {
         match_each_integer_ptype!(self.ptype(), |$P| {
-            let min: $P = self.reference().try_into().unwrap();
-            let shifted_min = min >> self.shift();
-            let unwrapped_value: $P = value.cast(self.dtype())?.try_into().unwrap();
-            let shifted_value: $P = unwrapped_value >> self.shift();
-            // Make sure that smaller values are still smaller and not larger than (which they would be after wrapping_sub)
-            if shifted_value < shifted_min {
-                return Ok(SearchResult::NotFound(0));
-            }
-
-            let translated_scalar = Scalar::primitive(
-                shifted_value.wrapping_sub(shifted_min),
-                value.dtype().nullability(),
-            )
-            .reinterpret_cast(self.ptype().to_unsigned());
-            search_sorted(&self.encoded(), translated_scalar, side)
+            search_sorted_typed::<$P>(self, value, side)
         })
     }
 }
 
+fn search_sorted_typed<T>(
+    array: &FoRArray,
+    value: &Scalar,
+    side: SearchSortedSide,
+) -> VortexResult<SearchResult>
+where
+    T: NativePType
+        + for<'a> TryFrom<&'a Scalar, Error = VortexError>
+        + Shr<u8, Output = T>
+        + Shl<u8, Output = T>
+        + WrappingSub
+        + WrappingAdd
+        + AddAssign
+        + Into<PValue>,
+{
+    let min: T = array.reference().try_into()?;
+    let primitive_value: T = value.cast(array.dtype())?.as_ref().try_into()?;
+    // Make sure that smaller values are still smaller and not larger than (which they would be after wrapping_sub)
+    if primitive_value < min {
+        return Ok(SearchResult::NotFound(0));
+    }
+
+    // When the values in the array are shifted, not all values in the domain are representable in the compressed
+    // space. Multiple different search values can translate to same value in the compressed space.
+    //
+    // For values that are not representable in the compressed array we know they wouldn't be found in the array
+    // in order to find index they would be inserted at we search for next value in the compressed space
+    let mut translated_value = primitive_value.wrapping_sub(&min) >> array.shift();
+    let mut non_representable = false;
+    if (translated_value << array.shift()).wrapping_add(&min) != primitive_value {
+        translated_value += T::from(1).unwrap();
+        non_representable = true;
+    }
+
+    let translated_scalar = Scalar::primitive(translated_value, value.dtype().nullability())
+        .reinterpret_cast(array.ptype().to_unsigned());
+
+    Ok(
+        match search_sorted(&array.encoded(), translated_scalar, side)? {
+            SearchResult::Found(i) => {
+                if non_representable {
+                    // If we are searching from the right and our value has not been representable we might have hit
+                    // the next value and need to shift the result
```

Comment:
this intuitively makes sense but would love eventually to have some more rigorous math proof